### PR TITLE
Feature/lifecycle

### LINKED ISSUE #1053: Add lifecycle support to applications
To support better handling of low-power background operations, or simply pausing games etc when focus is lost we should add some lifecycle handling.

I think that this should handle:
* OnBackground (or OnWillBackground?)
* OnForground (or OnDidForeground?)

I feel like more should be added, but cannot think just now. Apps can already handle DidStart and WillStop by doing:
```go
func main() {
    a := app.New()
    DidStart()
    a.ShowAndRun()
    WillStop()
}
```

### PR DESCRIPTION
**Pull request by andydotxyz**
### Description:
Simple lifecycle API for start/stop and focus gained/lost.

Question: Should it be OnForegrounded & OnBackgrounded instead (just felt too wordy...)

Fixes #1053

### Checklist:

- [ ] Tests included.
- [x] Lint and formatter run with no errors.
- [x] Tests all pass.
- [x] Public APIs match existing style.

### CODE REVIEW

**Comment by andydotxyz** [comment]
Maybe OnEnteredBackground and OnEnteredForeground?

**Review by stuartmscott — APPROVED** [review]
LGTM

In future we may want to add `SetOnLowMemory` so devs can free some resources, fyne could also purge the texture cache.

**Comment by andydotxyz** [comment]
Agreed. I am thinking about something for timezone/DST changes as well
